src/v/security/tests/mtls_test.cc

```diff
@@ -0,0 +1,135 @@
+  "default",
+  "DEFAUL",
+  "DEFAULT/L",
+  "DEFAULT/U",
+  "RULE:CN=(.*?),OU=ServiceUsers.*/$1",
+  "rule:^CN=(.*?),OU=ServiceUsers.*$/$1/",
+  "RULE:^CN=(.*?),OU=ServiceUsers.*$/$1/L/U",
+  "RULE:^CN=(.*?),OU=ServiceUsers.*$/L",
+  "RULE:^CN=(.*?),OU=ServiceUsers.*$/U",
+  "RULE:^CN=(.*?),OU=ServiceUsers.*$/LU"};
+
+BOOST_DATA_TEST_CASE(
+  test_mtls_invalid_rules, bdata::make(mtls_invalid_rules), c) {
+    BOOST_REQUIRE_THROW(principal_mapper{c}, std::runtime_error);
+}
+
+struct record {
+    record(std::string_view e, std::string_view i)
+      : expected{e}
+      , input{i} {}
+    std::string_view expected;
+    std::string_view input;
+    friend std::ostream& operator<<(std::ostream& os, const record& r) {
+        fmt::print(os, "input: `{}`, expected: `{}`", r.input, r.expected);
+        return os;
+    }
+};
+
+static std::array<record, 5> mtls_principal_mapper_data{
+  record{"duke", "CN=Duke,OU=ServiceUsers,O=Org,C=US"},
+  {"duke@sme", "CN=Duke,OU=SME,O=mycp,L=Fulton,ST=MD,C=US"},
+  {"DUKE@SME", "cn=duke,ou=sme,dc=mycp,dc=com"},
+  {"DUKE", "cN=duke,OU=JavaSoft,O=Sun Microsystems"},
+  {"OU=JavaSoft,O=Sun Microsystems,C=US",
+   "OU=JavaSoft,O=Sun Microsystems,C=US"}};
+
+BOOST_DATA_TEST_CASE(
+  test_mtls_principal_mapper, bdata::make(mtls_principal_mapper_data), c) {
+    security::tls::principal_mapper mapper{
+      "RULE:^CN=(.*?),OU=ServiceUsers.*$/$1/L, "
+      "RULE:^CN=(.*?),OU=(.*?),O=(.*?),L=(.*?),ST=(.*?),C=(.*?)$/$1@$2/L, "
+      "RULE:^cn=(.*?),ou=(.*?),dc=(.*?),dc=(.*?)$/$1@$2/U, "
+      "RULE:^.*[Cc][Nn]=([a-zA-Z0-9.]*).*$/$1/U, "
+      "DEFAULT"};
+    BOOST_REQUIRE_EQUAL(c.expected, *mapper.apply(c.input));
+}
+
+static std::array<record, 17> mtls_rule_splitting_data{
+  record{"[]", ""},
+  {"[DEFAULT]", "DEFAULT"},
+  {"[RULE:/]", "RULE://"},
+  {"[RULE:/.*]", "RULE:/.*/"},
+  {"[RULE:/.*/L]", "RULE:/.*/L"},
+  {"[RULE:/, DEFAULT]", "RULE://,DEFAULT"},
+  {"[RULE:/, DEFAULT]", "  RULE:// ,  DEFAULT  "},
+  {"[RULE:   /     , DEFAULT]", "  RULE:   /     / ,  DEFAULT  "},
+  {"[RULE:  /     /U, DEFAULT]", "  RULE:  /     /U   ,DEFAULT  "},
+  {"[RULE:([A-Z]*)/$1/U, RULE:([a-z]+)/$1, DEFAULT]",
+   "  RULE:([A-Z]*)/$1/U   ,RULE:([a-z]+)/$1/,   DEFAULT  "},
+  {"[]", ",   , , ,      , , ,   "},
+  {"[RULE:/, DEFAULT]", ",,RULE://,,,DEFAULT,,"},
+  {"[RULE: /   , DEFAULT]", ",  , RULE: /   /    ,,,   DEFAULT, ,   "},
+  {"[RULE:   /  /U, DEFAULT]", "     ,  , RULE:   /  /U    ,,  ,DEFAULT, ,"},
+  {R"([RULE:\/\\\(\)\n\t/\/\/])", R"(RULE:\/\\\(\)\n\t/\/\//)"},
+  {R"([RULE:\**\/+/*/L, RULE:\/*\**/**])",
+   R"(RULE:\**\/+/*/L,RULE:\/*\**/**/)"},
+  {"[RULE:,RULE:,/,RULE:,\\//U, RULE:,/RULE:,, RULE:,RULE:,/L,RULE:,/L, RULE:, "
+   "DEFAULT, /DEFAULT, DEFAULT]",
+   "RULE:,RULE:,/,RULE:,\\//U,RULE:,/RULE:,/,RULE:,RULE:,/L,RULE:,/L,RULE:, "
+   "DEFAULT, /DEFAULT/,DEFAULT"},
+};
+BOOST_DATA_TEST_CASE(
+  test_mtls_rule_splitting, bdata::make(mtls_rule_splitting_data), c) {
```

Comment:
What if we used YAML arrays in the config instead of a comma separated string?  I slightly worry about having to document comma escaping rules for users (although equally getting people to write syntactically correct yaml arrays isn't always easy either)

Reply:
Existing practice: https://github.com/apache/kafka/blob/trunk/clients/src/test/java/org/apache/kafka/common/security/ssl/SslPrincipalMapperTest.java